File: QuickCar/Core/Care/ChallengeEngine.swift
import Foundation
import SwiftData

@MainActor
final class ChallengeEngine: CareFeature {
    let id = "challenges"
    let requiredPIDs: Set<UInt8> = []
    let optionalPIDs: Set<UInt8> = []

    private let modelContext: ModelContext
    private let coldShield: ColdEngineShield

    struct Definition {
        let key: String
        let difficulty: String
        let points: Int
        let defaultTarget: Double
    }

    static let pool: [Definition] = [
        .init(key: "cleanWarmup", difficulty: "easy", points: 30, defaultTarget: 5),
        .init(key: "steadyCruise", difficulty: "easy", points: 30, defaultTarget: 5),
        .init(key: "idleDiet", difficulty: "medium", points: 50, defaultTarget: 0.7),
        .init(key: "softFoot", difficulty: "medium", points: 50, defaultTarget: 3),
        .init(key: "efficiencyJump", difficulty: "medium", points: 50, defaultTarget: 0.95),
        .init(key: "zeroHarshBrake", difficulty: "hard", points: 80, defaultTarget: 0),
        .init(key: "thermalDiscipline", difficulty: "hard", points: 80, defaultTarget: 0),
        .init(key: "nineties", difficulty: "hard", points: 80, defaultTarget: 5),
        .init(key: "morningMaster", difficulty: "hard", points: 80, defaultTarget: 5)
    ]

    init(modelContext: ModelContext, coldShield: ColdEngineShield) {
        self.modelContext = modelContext
        self.coldShield = coldShield
    }

    func isEnabled(settings: AppSettings) -> Bool { settings.careWeeklyChallenges }

    func evaluate(snapshot: VehicleSnapshot, context: inout CareContext) -> [CareCue] {
        ensureWeekChallenges(now: context.now)
        return []
    }

    func onTripEnded(trip: Trip, context: CareContext) -> [CareCue] {
        ensureWeekChallenges(now: context.now)
        let week = Self.weekStart(for: context.now)
        let rows = currentWeek(weekStart: week)
        for row in rows where row.completedAt == nil {
            switch row.challengeKey {
            case "cleanWarmup":
                if coldShield.coldViolationsThisTrip == 0 {
                    row.current += 1
                }
            case "zeroHarshBrake":
                let harsh = (trip.events ?? []).filter { $0.typeRaw == "harshBrake" }.count
                if harsh == 0, trip.distanceKm > 0 {
                    row.current = min(row.target, row.current + trip.distanceKm)
                } else if harsh > 0 {
                    row.current = -1 // failed
                }
            case "thermalDiscipline":
                // hot shutdowns counted elsewhere; treat as pass if no hotShutdown events today
                break
            case "nineties":
                if (trip.scoreTotal ?? 0) >= 90 {
                    row.current += 1
                } else {
                    row.current = 0
                }
            case "softFoot":
                let harsh = (trip.events ?? []).filter { $0.typeRaw == "harshAccel" }.count
                if harsh == 0 {
                    row.current = min(row.target, row.current + trip.distanceKm)
                }
            default:
                break
            }
            if row.current >= row.target, row.current >= 0 {
                row.completedAt = context.now
            }
        }
        try? modelContext.save()
        return []
    }

    func currentWeek(weekStart: Date) -> [ChallengeProgress] {
        let all = (try? modelContext.fetch(FetchDescriptor<ChallengeProgress>())) ?? []
        return all.filter { Calendar.current.isDate($0.weekStart, inSameDayAs: weekStart) }
    }

    func ensureWeekChallenges(now: Date) {
        let week = Self.weekStart(for: now)
        let existing = currentWeek(weekStart: week)
        guard existing.isEmpty else { return }

        let recentKeys: Set<String> = {
            let all = (try? modelContext.fetch(FetchDescriptor<ChallengeProgress>())) ?? []
            let cutoff = Calendar.current.date(byAdding: .day, value: -21, to: now) ?? now
            return Set(all.filter { $0.weekStart >= cutoff }.map(\.challengeKey))
        }()

        func pick(difficulty: String) -> Definition {
            let candidates = Self.pool.filter { $0.difficulty == difficulty && !recentKeys.contains($0.key) }
            return candidates.randomElement()
                ?? Self.pool.filter { $0.difficulty == difficulty }.randomElement()!
        }
        for def in [pick(difficulty: "easy"), pick(difficulty: "medium"), pick(difficulty: "hard")] {
            modelContext.insert(ChallengeProgress(
                challengeKey: def.key,
                weekStart: week,
                target: def.defaultTarget,
                difficultyRaw: def.difficulty,
                points: def.points
            ))
        }
        try? modelContext.save()
    }

    static func weekStart(for date: Date) -> Date {
        var cal = Calendar.current
        cal.firstWeekday = 2 // Monday
        let comps = cal.dateComponents([.yearForWeekOfYear, .weekOfYear], from: date)
        return cal.date(from: comps) ?? date
    }
}

File: mobile/src/core/care/challengeEngine.ts
import { and, eq } from "drizzle-orm";
import { db } from "../storage/db";
import { challengeProgress } from "../storage/schema";
import { coldEngineShield } from "./coldEngineShield";
import type { VehicleSnapshot } from "../obd/vehicleSnapshot";
import type { Trip } from "../storage/models";
import type { CareCue, CareContext } from "./careTypes";
import type { CareFeature } from "./careFeature";

interface Definition {
  key: string;
  difficulty: "easy" | "medium" | "hard";
  points: number;
  defaultTarget: number;
}

const POOL: Definition[] = [
  { key: "cleanWarmup", difficulty: "easy", points: 30, defaultTarget: 5 },
  { key: "steadyCruise", difficulty: "easy", points: 30, defaultTarget: 5 },
  { key: "idleDiet", difficulty: "medium", points: 50, defaultTarget: 0.7 },
  { key: "softFoot", difficulty: "medium", points: 50, defaultTarget: 3 },
  { key: "efficiencyJump", difficulty: "medium", points: 50, defaultTarget: 0.95 },
  { key: "zeroHarshBrake", difficulty: "hard", points: 80, defaultTarget: 0 },
  { key: "nineties", difficulty: "hard", points: 80, defaultTarget: 5 },
];

/** Monday-start ISO week bucket. */
export function weekStart(date: number): number {
  const d = new Date(date);
  const day = d.getDay();
  const diff = (day === 0 ? -6 : 1) - day;
  d.setHours(0, 0, 0, 0);
  d.setDate(d.getDate() + diff);
  return d.getTime();
}

class ChallengeEngine implements CareFeature {
  id = "challenges";

  isEnabled(settings: { careWeeklyChallenges: boolean }): boolean {
    return settings.careWeeklyChallenges;
  }

  evaluate(_snapshot: VehicleSnapshot, context: CareContext): CareCue[] {
    this.ensureWeekChallenges(context.now).catch(() => undefined);
    return [];
  }

  async currentWeek(week: number) {
    const all = await db.select().from(challengeProgress);
    return all.filter((r) => r.weekStart === week);
  }

  async ensureWeekChallenges(now: number): Promise<void> {
    const week = weekStart(now);
    const existing = await this.currentWeek(week);
    if (existing.length > 0) return;

    const all = await db.select().from(challengeProgress);
    const cutoff = now - 21 * 86_400_000;
    const recentKeys = new Set(all.filter((r) => r.weekStart >= cutoff).map((r) => r.challengeKey));

    const pick = (difficulty: Definition["difficulty"]) => {
      const candidates = POOL.filter((d) => d.difficulty === difficulty && !recentKeys.has(d.key));
      const pool = candidates.length ? candidates : POOL.filter((d) => d.difficulty === difficulty);
      return pool[Math.floor(Math.random() * pool.length)]!;
    };

    for (const def of [pick("easy"), pick("medium"), pick("hard")]) {
      await db.insert(challengeProgress).values({
        id: `chal_${week}_${def.key}`,
        challengeKey: def.key,
        weekStart: week,
        target: def.defaultTarget,
        difficulty: def.difficulty,
        points: def.points,
      });
    }
  }

  async onTripEnded(trip: Trip, context: CareContext): Promise<CareCue[]> {
    await this.ensureWeekChallenges(context.now);
    const week = weekStart(context.now);
    const rows = await this.currentWeek(week);

    for (const row of rows) {
      if (row.completedAt != null) continue;
      let current = row.current;
      switch (row.challengeKey) {
        case "cleanWarmup":
          if (coldEngineShield.coldViolationsThisTrip === 0) current += 1;
          break;
        case "zeroHarshBrake": {
          const harsh = (trip.events ?? []).filter((e) => e.type === "harshBrake").length;
          current = harsh === 0 && trip.distanceKm > 0 ? Math.min(row.target, current + trip.distanceKm) : -1;
          break;
        }
        case "nineties":
          current = (trip.scoreTotal ?? 0) >= 90 ? current + 1 : 0;
          break;
        case "softFoot": {
          const harsh = (trip.events ?? []).filter((e) => e.type === "harshAccel").length;
          if (harsh === 0) current = Math.min(row.target, current + trip.distanceKm);
          break;
        }
        default:
          break;
      }
      const completedAt = current >= row.target && current >= 0 ? context.now : null;
      await db.update(challengeProgress).set({ current, completedAt }).where(eq(challengeProgress.id, row.id));
    }
    return [];
  }
}

export const challengeEngine = new ChallengeEngine();
